### PR TITLE
[docs] fix benders_decomposition.jl

### PR DESCRIPTION
**Pull request by odow**
Closes https://github.com/jump-dev/JuMP.jl/issues/3531

Thanks for catching this @WalterMadelim. I'm surprised that it didn't error!

### CODE REVIEW

File: docs/src/tutorials/algorithms/benders_decomposition.jl [hunk 1]
```diff
@@ -266,7 +266,7 @@ function my_callback(cb_data)
         return
     end
     cut = @build_constraint(θ >= ret.obj + -ret.π' * A_1 * (x .- x_k))
-    MOI.submit(model, MOI.LazyConstraint(cb_data), cut)
+    MOI.submit(lazy_model, MOI.LazyConstraint(cb_data), cut)
```
Comment (odow) [inline docs/src/tutorials/algorithms/benders_decomposition.jl:269]: I should look at why this didn't error. It should have...

Comment (WalterMadelim) [reply docs/src/tutorials/algorithms/benders_decomposition.jl:269]: image
It has indeed, after changing GLPK to Gurobi. (I have not installed GLPK)

Comment (odow) [reply docs/src/tutorials/algorithms/benders_decomposition.jl:269]: Ah, so this might be an issue with GLPK, not with JuMP. Thanks for looking.

**Comment by odow** [comment]
Now I don't understand what's going on here:
image

**Comment by odow** [comment]
GLPK callbacks are weird

**Comment by WalterMadelim** [comment]
I think the output in your figure is normal. There are multiple "Terminating with the optimal solution", because this is a callback function, which will be called frequently during solving, and the solver's terminating criterion doesn't depend on user's callback function. As for the Jumping up - Jumping down lower bound, this is also reasonable because the problem is changing during solving. The puzzling point is why GLPK can return the right answer, even if you submit to `model` not `lazy_model`.

But it should be fine after revising. I agree with the new callback function.

**Comment by WalterMadelim** [comment]
> GLPK callbacks are weird

Actually I would prefer to use a solver-dependent callback. (For example, I would refer to Gurobi.jl directly when I have a need)

File: docs/src/tutorials/algorithms/benders_decomposition.jl [hunk 2]
```diff
@@ -239,34 +239,27 @@ y_optimal = optimal_ret.y
...
-    global k += 1
+    status = callback_node_status(cb_data, lazy_model)
+    if status != MOI.CALLBACK_NODE_STATUS_INTEGER
+        ## Only add the constraint if `x` is an integer feasible solution
+        return
+    end
     x_k = callback_value.(cb_data, x)
     θ_k = callback_value(cb_data, θ)
-    lower_bound = c_1' * x_k + θ_k
+    number_of_subproblem_solves += 1
```
Comment (odow) [inline docs/src/tutorials/algorithms/benders_decomposition.jl:256]:
```suggestion
    global number_of_subproblem_solves += 1
```

**Comment by odow** [comment]
I don't understand why GLPK calls so many times: 
https://jump.dev/JuMP.jl/previews/PR3532/tutorials/algorithms/benders_decomposition/#Callback-method

image

**Comment by WalterMadelim** [comment]
I thought there might be no one who is more familiar with this than you, because it seems that you are the main contributer of GLPK.jl.🥲

From my point of view, the `function my_callback(cb_data)` is special, because it is controlled by the solver rather than Julia. I tried to put break points `@bp`'s inside it, but they were never hitted. Therefore this function is more likely to be a component of the branch and cut module: when the main algorithm need a user's callback, then a lazy cut is retrieved and stored.

Therefore, I think the name `number_of_subproblem_solves` is not proper. It's the number of times that callback module in GLPK is executed when `status == MOI.CALLBACK_NODE_STATUS_INTEGER`. Subproblems modeling and branching should reside in a deeper module than callback module in GLPK (my comprehension).

This adapted version of TSP callback code (using MOI without JuMP) might be a comparison. Solver is Gurobi, the problem is larger and the counter reaches `20`, compared with your result `341`. (you can try it)

https://github.com/WalterMadelim/p8f.jl/blob/main/src/TSP_GRB_callback.jl

**Comment by odow** [comment]
> I thought there might be no one who is more familiar with this than you, because it seems that you are the main contributer of GLPK.jl

Yes :smile: I meant this as "given I understand exactly how it works, it is confusing that it is called so many times."

File: docs/src/tutorials/algorithms/benders_decomposition.jl [hunk 3]
```diff
@@ -276,8 +269,12 @@ set_attribute(lazy_model, MOI.LazyConstraintCallback(), my_callback)
 
 optimize!(lazy_model)
 
-# Note how this problem also takes 4 iterations to converge, but the sequence
-# of bounds is different compared to the iterative method.
+# For this model, the callback algorithm required more solves of the subproblem:
+
+number_of_subproblem_solves
+
+# But for larger problem, you cann expect the callback algorithm to be more
```
Comment (odow) [inline docs/src/tutorials/algorithms/benders_decomposition.jl:276]:
```suggestion
# But for larger problem, you can expect the callback algorithm to be more
```

File: docs/src/tutorials/algorithms/benders_decomposition.jl [hunk 4]
```diff
@@ -276,8 +269,12 @@ set_attribute(lazy_model, MOI.LazyConstraintCallback(), my_callback)
 
 optimize!(lazy_model)
 
-# Note how this problem also takes 4 iterations to converge, but the sequence
-# of bounds is different compared to the iterative method.
+# For this model, the callback algorithm required more solves of the subproblem:
+
+number_of_subproblem_solves
+
+# But for larger problem, you can expect the callback algorithm to be more
```
Comment (odow) [inline docs/src/tutorials/algorithms/benders_decomposition.jl:276]:
```suggestion
# But for larger problems, you can expect the callback algorithm to be more
```

**Pull request merged**